Phpstan level 3

Art4 opened the pull request:
This PR fixes all 156 errors that are reported by PHPStan level 3. I've split addon configuration from the `phpstan.neon` file into a new `phpstan-addons.neon` file. This way we can run the phpstan jobs for friendica and addons separatly. This way I can work in parallel on the PHPStan errors in both files.

In the long run the `phpstan-addons.neon` file should be move into the addons repository. This should be done once we have setup woodpecker in the addons repository.

This PR targets the 2025.05 release.

- [x] Requires merge of https://git.friendi.ca/friendica/friendica-addons/pulls/1603

### Changelog

Notes for the `CHANGELOG.md` from #14722:

```md
### Deprecated

- The method `\Friendica\BaseRepository::_selectOne()` is deprecated, use `\Friendica\BaseRepository::_selectFirstRowAsArray()` instead.
```

Art4 committed:
Run PHPStan with level 3

Art4 committed:
Fix 11 errors

Art4 committed:
Fix 2 PHPStan errors

Art4 committed:
Fix PHPStan error in Security class

Art4 committed:
Refactory PermissionSet Repository, add BaseRepository::_selectFirstRowAsArray()

Art4 committed:
Deprecate BaseRepository::_selectOne()

Art4 committed:
Refactor DisporaContact Repository, fix PHPStan error

Art4 committed:
fix 17 PHPStan errors

Art4 committed:
Fix 27 PHPStan errors

Art4 committed:
Fix 11 PHPStan errors

Art4 committed:
Split phpstan.neon files for addon checks

Art4 committed:
Add CI step for PHPStan in addons folder

Art4 committed:
Fix 4 PHPStan errors

Art4 committed:
Fix 5 PHPStan errors

Art4 committed:
Deprecate BaseCollection::map() method

Art4 committed:
Fix 10 PHPStan errors

Art4 committed:
Fix 2 PHPStan errors

MrPetovan commented on src/BaseCollection.php:
```diff
@@ -83,12 +83,16 @@ public function column(string $column, $index_key = null): array
 	 */
 	public function map(callable $callback): BaseCollection
 	{
+		@trigger_error('`' . __METHOD__ . '()` is deprecated since 2025.05 and will be removed after 5 months, use `array_map()` instead.', E_USER_DEPRECATED);
```
`array_map` returns a regular array, not a `BaseCollection` object, what is the purpose of this deprecation?

MrPetovan replied:
Now that I have read the whole patch, I get that you moved that code to each inherited method, but now I'm confused about the duplicated code. This was working fine, why change it?

Art4 replied:
The problem with `BaseCollection::map()` is that it returns a `BaseCollection` object, but every method that uses this `map()` returns a childclass like `Notifications`. So we need to check in every method if `BaseCollection::map()` really returns an instance of `Notifications` (or another class):

```diff
	public function setSeen(): Notifications
	{
		// $notifications is instance of BaseCollection, but we want to return Notifications
		$notifications = $this->map(function (Entity\Notification $Notification) {
			$Notification->setSeen();
		});
+
+		if (!$notifications instanceof Notifications)
+		{
+			// What shall we do now?
+			throw new \Exception();
+		}

		return $notifications;
	}
```

That's why I decided to not call `BaseCollection::map()` at all and I updated all child classes. Now no class uses `BaseCollection::map()` anymore. Because of BC I deprecated this method instead of deleting it.

Sure, `array_map()` returns an array, but deprecation notes are not meant as a drop-in replacement, but as a suggestion. One has to think about the replacement before changing the code.

MrPetovan replied:
I wasn't able to specific the `static` return type, so that's why I had to go with the generic `BaseCollection`. But this is only a type-hinting problem, in practice we always instantiate the child class with `$class = get_class($this)` that you have had to duplicate as well and this concern is moot.

Art4 replied:
> I wasn't able to specific the `static` return type, so that's why I had to go with the generic `BaseCollection`. But this is only a type-hinting problem, 

This code is error prone because atm we have to "believe" that `BaseCollection::map()` returns the correct implementation. That's what PHPStan is complaining about. We should either check the returned value (with `instanceof` as shown above) or implement the `map()` functionality inside the child classes ourself (that's my recommendation).

> in practice we always instantiate the child class with `$class = get_class($this)` that you have had to duplicate as well and this concern is moot.

I would also like to get rid of the `$class = get_class($this);` and `return new $class(` code, but I have to keep it for BC. Because the Notifications class is not final it's possible that someone has extended this class and using return `new Notifications(` would break their code.

MrPetovan replied:
I understand your reasoning but Friendica isn't a library that people would ever extend. In practice we control 99% of the code that depends on Friendica and the remaining 1% are unofficial addons with little use, otherwise we're prompt to make them official and assume maintenance.

This is why I'm weary of automated tools, they lack context, and while they are technically correct, acting on these complaints would be costly for absolutely no return given our situation. In this present case, this would lead to code duplication (since you're still using `get_class()`) for no practical benefit since only we are creating child classes. But at least PHPStan would be quieter? This makes little sense to me.

That said, I'm curious if you know of a PHP-7.4-compatible way of creating this kind of more strongly typed model and collections structure without having to copy the boilerplate code every time we create a new one (which doesn't happen often, but still). This was meant as a stepping stone from anonymous arrays widely used throughout the code but it isn't meant to be end end-all-be-all. In particular, I'm not sure how to approach model dependencies (contact depends on gserver which depends on...)

annando replied:
I haven't read the whole thread (it is a recurring theme for me), but I want to support @MrPetovan here. We are in control of both the core and most of the addons. We can allow breaking changes. In fact we did so many times and got now complaints as far as I can remember. I really would like to avoid any introduced complication here, just because someone maybe could have used something that we changed now.

Art4 replied:
> I understand your reasoning but Friendica isn't a library that people would ever extend. In practice we control 99% of the code that depends on Friendica and the remaining 1% are unofficial addons with little use, otherwise we're prompt to make them official and assume maintenance.

That's great to hear. Then I would recommend to change the [BC promise](https://github.com/friendica/friendica/blob/develop/doc/Developers-Intro.md#backward-compatibility-promise) exceptions from "extending final classes" to "extending any non-abstract classes"? Alternative I can replace the `get_class()` call directly and note the breaking change. What do you think? (Update: see #14849)

> This is why I'm weary of automated tools, they lack context, and while they are technically correct, acting on these complaints would be costly for absolutely no return given our situation. In this present case, this would lead to code duplication (since you're still using get_class()) for no practical benefit since only we are creating child classes. But at least PHPStan would be quieter? This makes little sense to me.

This sounds like we have a different prioritization. Goals of static code analysis are improving the code quality, reducing technical dept and more. Benefits are enhancing reliability, clean and maintainable code, early detection of bugs/vulnerabilities and more. You can try to estimate how many warnings in logs the fixes will avoid already. For me saving 2 lines of code or DRY is secondary (we [talked already about WET](https://github.com/friendica/friendica/pull/14559#issuecomment-2510220874)). This changes are in no way costly but priceless. They show us where the next problems lies we need to solve.

> That said, I'm curious if you know of a PHP-7.4-compatible way of creating this kind of more strongly typed model and collections structure without having to copy the boilerplate code every time we create a new one (which doesn't happen often, but still). 

Sure: We should use composition over inheritance. But we are not yet at this point to refactor this.

> This was meant as a stepping stone from anonymous arrays widely used throughout the code but it isn't meant to be end end-all-be-all. In particular, I'm not sure how to approach model dependencies (contact depends on gserver which depends on...)

The widely use of arrays is [another problem](https://refactoring.guru/smells/primitive-obsession) that will be tackled later too. I already have an idea how to approach this but again we are not yet at this point to refactor this problem.

MrPetovan replied:
> Sure: We should use composition over inheritance. But we are not yet at this point to refactor this.

Isn't this the point of the Repositories that use the provided Factory to instantiate Models in the `_select` method that you also deprecated it in this PR? I may also have a completely wrong understanding of composition.

> This changes are in no way costly but priceless. They show us where the next problems lies we need to solve.

I'm not sure I see it. I'm on board for the next refactoring scheme, but `BaseCollection` is far from WET, there are 21 child classes that could use the `map()` method that currently works exactly as intended in our context. This means that if we every add a `map()` usage for a class that wasn't using it before, then the log will scream bloody murder until we just copy the `BaseCollection` method code into its dedicated child class method? Or worst, use `array_map()` with extra steps to care for the object aspect? This sounds brittle to me, when using `ChildClass->map()` was already available and doing what one would expect it to do.

Can we skip this PHPStan warning for now?

Art4 replied:
> Isn't this the point of the Repositories that use the provided Factory to instantiate Models in the `_select` method that you also deprecated it in this PR?

That's the point: The child classes already have access to `$db` and `$factory`, but they give them to `BaseRepository` and they have to call `BaseRepository::_select()` only to get a collection in a wrong type. So now we have to make sure that `BaseRepository::_select()` really returns the correct type (this is what PHPStan warns about). instead we should make use of the  `$db` and `$factory` in the child classes directly and could probably remove the `BaseRepository` class.

> I'm not sure I see it. I'm on board for the next refactoring scheme, but `BaseCollection` is far from WET, there are 21 child classes that could use the `map()` method that currently works exactly as intended in our context. This means that if we every add a `map()` usage ...

And that's my next point: There are 21 child classes of `BaseCollection`, but the `map()` method is only called in 3 child classes (where it is overridden in 1 class). That means that 18 child classes have a `map()` method they don't need. So my advise is **not** blindly copy the `map()` method from the `BaseCollection` class into the child classes, but only where it is needed. And after we have removed the inheritance to the `BaseCollection` class most of the (old child) classes will become smaller.

And now comes the WET and "composition over inheritance"-part: If we see that the remaining `map()` methods could be abstracted, than we can move this method into the `Factory` or create even a new service like `CollectionMapper` and inject them only into this classes, or something else. But we cannot decide this yet.

> Can we skip this PHPStan warning for now?

We could make the possible error explicitly so everyone (and PHPStan) can see that we are aware and take care of a possible type mismatch.

```diff
	public function setSeen(): Notifications
	{
		// $notifications is instance of BaseCollection, but we want to return Notifications
		$notifications = $this->map(function (Entity\Notification $Notification) {
			$Notification->setSeen();
		});
+
+		if (!$notifications instanceof Notifications)
+		{
+			// Show the possible error explicitly
+			throw new \Exception(sprintf(
+				'BaseCollection::map()' should return instance of %s, but returns %s instead.',
+				Notifications::class,
+				get_class($notifications)
+			);
+		}

		return $notifications;
	}
```

MrPetovan replied:
So this check would only be present where we actually use the method? I can live with that.

> And that's my next point: There are 21 child classes of BaseCollection, but the map() method is only called in 3 child classes (where it is overridden in 1 class). That means that 18 child classes have a map() method they don't need.

But they could need in the future, we aren't done trying to get rid of anonymous arrays. That's my worry: that we may need it in the future and have to rewrite it simply because it wouldn't be accessible by default through inheritance. Again, this is already working as expected, and the check you are intending to add will never be hit if we don't change the `BaseCollection->map` method, which is small and contained enough to never need a functional do-over.

The PHPStan warning is there because in PHP7.4 we can't type-hint return value as `static`, but it isn't because it would ever fail as it is. To me the refactoring you're suggesting would be like setting the table with just the cutlery individual people need for their specific meal. This isn't how I set the table, I always add a fork, a knife and a spoon, no matter what we might be having for a meal. And I don't expect the fork, the knife or the spoon to ever change form. But PHPStan, that has never eaten a meal in its life, might object on general grounds. This is fine, but it doesn't mean we have to listen to it.

Art4 replied:
> So this check would only be present where we actually use the method? I can live with that.

Yes, that's correct. 👍 I'll change the PR to throw the explicit exception instead of deprecating the `map()` method.

> But they could need in the future, we aren't done trying to get rid of anonymous arrays. That's my worry: that we may need it in the future and have to rewrite it simply because it wouldn't be accessible by default through inheritance. Again, this is already working as expected, and the check you are intending to add will never be hit if we don't change the `BaseCollection->map` method, which is small and contained enough to never need a functional do-over.

Getting rid of the arrays is a separate task I'd like to tackle in the future. And about your worry: [YAGNI](https://dev.to/alisamir/yagni-you-arent-gonna-need-it-a-key-principle-in-software-development-167m). Adding functionality later is cheaper than adding it upfront because (right now) we have to deal with this (unused) code, have to refactor it, have to write test for it and so on. This leads to complex code without any need. We should get rid of it and only add code if we need it.

> The PHPStan warning is there because in PHP7.4 we can't type-hint return value as `static`, but it isn't because it would ever fail as it is.

I'm not sure how using `static` would help us? Full return type covariance was added with PHP 7.4, see https://www.php.net/manual/en/language.oop5.variance.php 

I have reproduced the error here: https://phpstan.org/r/7d335447-bf34-4668-8883-9fc71b60f81f

> To me the refactoring you're suggesting would be like setting the table with just the cutlery individual people need for their specific meal. This isn't how I set the table, I always add a fork, a knife and a spoon, no matter what we might be having for a meal. And I don't expect the fork, the knife or the spoon to ever change form. 

To stick with your analogy: We know very well what there is to eat because we have all the recipes in front of us. If we don't have recipes for soup, we never need to put soup spoons and soup plates on the table. And if we do add a soup recipe, only then is it necessary to provide soup plates and soup spoons. Before that, we put them on the table pointlessly, have to wash them again without using them and take care of them, even though they are never used. And we know that.

> But PHPStan, that has never eaten a meal in its life, might object on general grounds. This is fine, but it doesn't mean we have to listen to it.

I think you're misunderstandin PHPStan. 😊 The refactoring is mine, not PHPStan's. PHPStan is just saying something is wrong here. We have to take care of the solution ourselves.

While developing/refactoring I'm always focused on building clean, robust and testable code to reduce complexity. Less code contains less bugs, and allows us to move on faster again.

MrPetovan replied:
> I have reproduced the error here: https://phpstan.org/r/7d335447-bf34-4668-8883-9fc71b60f81f

That's a pretty cool playground I didn't know about. Your example didn't look anything like what we're doing to me so I went ahead and I put the actual `BaseCollection` code with the prospective `static` return type for `BaseCollection->map`: https://phpstan.org/r/3ba5aabf-31fc-4100-92b0-95e18d19f9b1

PHPStan returns the following error I don't understand:
> Method BaseCollection::map() should return static(BaseCollection) but returns BaseCollection.

What kind of type notation `static(BaseCollection)` is? I've never seen it before (and I've used TypeScript) and there's no specific documentation for errors, just recommendations to ignore them.

MrPetovan replied:
I've also tried with the following code:
```php
public function map(callable $callback): static
{
	return new static(array_map($callback, $this->getArrayCopy()), $this->getTotalCount());
}
```
And now it reports
> Unsafe usage of new static().

Which I do understand, including the ways to solve it.

Art4 replied:
I'm sorry, I mixed up two different problems. And about the missing static return in PHP 7.4 I understand now that you were referring to https://php.watch/versions/8.0/static-return-type

MrPetovan replied:
And I do understand now that your example applied to the separate `BaseRepository` issue that you fixed in the `GServer` class and for which I don't have a problem, the solution seems straightforward.

MrPetovan commented on src/BaseRepository.php:
```diff
@@ -130,18 +130,32 @@ protected function _select(array $condition, array $params = []): BaseCollection
+	}
+
+	/**
+	 * @deprecated 2025.05 Use `\Friendica\BaseRepository::_selectFirstRowAsArray()` instead
```
Same question, why moving back from structured objects back to anonymous arrays?

Art4 replied:
Same answer as with `BaseCollection::map()`: `BaseRepository::_selectOne()` return `BaseEntity` instance, but calling classes requires the entities as a specific childclass. So instead check the result for the correct class type, I select the first row as array and call the specific factory myself where I know that the result is the specific childclass.

```diff
// in src/Navigation/Notifications/Repository/Notification.php
	private function selectOne(array $condition, array $params = []): NotificationEntity
	{
-		return parent::_selectOne($condition, $params);
+		$fields = $this->_selectFirstRowAsArray( $condition, $params);
+
+		return $this->factory->createFromTableRow($fields);
	}
```

MrPetovan replied:
Same answer as well, the `static` return type isn't available, but in practice `_selectOne` always returns a child class object thanks to the `$this->factory` property which will never create a naked `BaseEntity` object.

MrPetovan commented on src/Module/Api/Mastodon/Search.php:
```diff
@@ -196,6 +196,11 @@ private function searchHashtags(string $q, bool $exclude_unreviewed, int $limit,
 		$tags = DBA::select('tag', ['name'], $condition, $params);
 
 		$hashtags = [];
+
+		if (!is_iterable($tags)) {
+			return $hashtags;
+		}
```
Please use `DBA::selectToArray` instead.
```suggestion
		$tags = DBA::selectToArray('tag', ['name'], $condition, $params);

		$hashtags = [];
```

Art4 replied:
Nice, thank you. 👍

MrPetovan commented on src/Module/Api/Twitter/Lists/Ownership.php:
```diff
@@ -48,8 +48,11 @@ protected function rawContent(array $request = [])
-		foreach ($circles as $circle) {
-			$lists[] = $this->friendicaCircle->createFromId($circle['id']);
+
+		if (is_iterable($circles)) {
```
Please use `Database->selectToArray` on line 47 instead of adding a condition.

Art4 replied:
Done.

MrPetovan commented on src/Navigation/Notifications/Collection/Notifications.php:
```diff
@@ -8,36 +8,50 @@
+		$class = get_class($this);
+
+		return new $class(
+			array_map(
+				function (NotificationEntity $notification) {
+					$notification->setSeen();
+				},
+				$this->getArrayCopy()
+			),
+			$this->getTotalCount(),
+		);
 	}
 
 	public function setDismissed(): Notifications
 	{
-		return $this->map(function (Entity\Notification $Notification) {
-			$Notification->setDismissed();
-		});
+		$class = get_class($this);
+
+		return new $class(
+			array_map(
+				function (NotificationEntity $notification) {
+					$notification->setDismissed();
+				},
+				$this->getArrayCopy(),
+			),
+			$this->getTotalCount(),
+		);
+
```
As much as I like cheeky multi-level indentation, I really think this isn't necessary and the inherited `BaseCollection->map` should do the trick, but I'm still curious to hear why you went through all this trouble.

Art4 replied:
I can create a new `mapNotifications()` method and refactor both methods to use this new method.

I would also like to get rid of the `$class = get_class($this);` and `return new $class(` code, but I have to keep it for BC. Because the `Notifications` class is not final it's possible that someone has extended this class and using `return new Notifications(` would break their code.

MrPetovan commented on src/Object/Api/Mastodon/Notification.php:
```diff
@@ -44,9 +44,9 @@ class Notification extends BaseDataTransferObject
 	protected $dismissed;
-	/** @var Account */
+	/** @var array */
 	protected $account;
-	/** @var Status|null */
+	/** @var array|null */
 	protected $status = null;
```
I think we should leave the strong types in, but update `BaseDataTransferObject->toArray()` to recursively call `toArray` on object properties instead of flatly using `get_object_vars()`.

Art4 replied:
I have only adapted the doctype to reflect reality.

I would recommend to move your suggestion into a separate PR. While refactoring I try to avoid such code changes.

MrPetovan replied:
Fair enough

MrPetovan commented on src/Util/HTTPSignature.php:
```diff
@@ -537,14 +538,12 @@ public static function fetchRaw(string $request, int $uid = 0, array $opts = [Ht
+		}
+
+		if (!$owner) {
+			throw new Exception('Could not find owner for uid ' . $uid);
```
This probably should log a critical error about the system account not being set up when `$uid` is empty.

Art4 replied:
I've checked every calling places. All places are calling it inside a try-catch block and are logging an exception like this:

```php
try {
	$curlResult = HTTPSignature::fetchRaw($request, $uid);
} catch (\Exception $exception) {
	DI::logger()->notice('Error fetching url', ['url' => $request, 'exception' => $exception]);
}
```

As one can see the `$exception` is always logged too, so throwing this new exception should be enough context.

Art4 committed:
Update src/Module/Api/Mastodon/Search.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
Use selectToArray() instead of select() and is_iterable()

Art4 committed:
Merge branch 'develop' into phpstan-level-3

Art4 committed:
Merge branch 'develop' into phpstan-level-3

Art4 committed:
Fix 7 PHPStan errors

Art4 committed:
fix 7 PHPStan errors in factoriy and Model

Art4 committed:
Handle missing attachment with correct response

Art4 committed:
Fix upload error handling for media

see https://github.com/friendica/friendica/pull/14475
refs #14475

Art4 committed:
Change 500 response to 422 response

see https://github.com/friendica/friendica/pull/14475

Art4 committed:
Refactor GServer Repository

Art4 committed:
simplify GServer repository

Art4 commented:
@MrPetovan Please take a look at the latest commits, where I refactored the GServer repository. First I checked all calls for the class and noticed, that only the `selectOneById()` method is used at all. The real work is done in the Database class. So I refactored the class in two steps:

1. In commit 38343ee417bd1cf4f37d76e13c59779451e8d12f I removed the extension of the `BaseRepository` class and moved the constructor injected parameter directly into the GServer repository class. I also extracted the code from `BaseRepository::_selectFirstRowAsArray()` directly inside the `GServer::selectOneById()` method.
2. Now in commit 767ac778c96eebc048a30a939b66d2e257cb1a82 I was able simplify the GServer repository class by removing the unnecessary dependency to the `LoggerInterface`. 

Now we have a nice and clean class with defined dependencies that returns defined types. In this case we don't need the overhead brought to us by the `BaseRepository`. If we need a new method we can create it. If we are finding ourself to implement the same method in a second or third place again, we can think about implementing it inside the `Database` class, creating a new interface for this functions or extract the code to a new class that would then injected to the places where it is needed.

Art4 committed:
Fix 2 PHPStan errors

Art4 committed:
Fix 13 PHPStan errors

Art4 committed:
remove deprecation of BaseCollection::map(), check return type of BaseCollection::map()

Art4 committed:
Merge branch 'develop' into phpstan-level-3

Art4 committed:
Fix 10 PHPStan errors

Art4 committed:
Deprecate BaseRepository::_selectOne()

set BaseRepository::_selectFirstRowAsArray() as final

Art4 committed:
Fix 10 PHPStan errors

Art4 committed:
Fix return type or UserSession::getUserIDForVisitorContactID()

Art4 committed:
Fix 6 PHPStan errors

Art4 committed:
ignore PHPStan error in deprecated code

Art4 committed:
fix 14 PHPStan errors in console

Art4 committed:
Fix possible false return type in PHP 7.4

Art4 committed:
Run phpstan in addons with level 3

Art4 commented:
This PR is ready for review, but requires the merge of https://git.friendi.ca/friendica/friendica-addons/pulls/1603

Please note that I didn't run the code style fixer for now to allow a better review. Once this PR is accepted I will run the fixer and add another commit.

MrPetovan commented on src/Contact/FriendSuggest/Repository/FriendSuggest.php:
```diff
@@ -49,20 +49,17 @@ private function convertToTableRow(FriendSuggestEntity $fsuggest): array
 	private function selectOne(array $condition, array $params = []): FriendSuggestEntity
 	{
-		return parent::_selectOne($condition, $params);
+		$fields = $this->_selectFirstRowAsArray( $condition, $params);
```
```suggestion
		$fields = $this->_selectFirstRowAsArray($condition, $params);
```

MrPetovan commented on src/Contact/LocalRelationship/Repository/LocalRelationship.php:
```diff
@@ -7,59 +7,54 @@
+	public function selectForUserContact(int $uid, int $cid): LocalRelationshipEntity
 	{
-		return $this->_selectOne(['uid' => $uid, 'cid' => $cid]);
+		$fields = $this->_selectFirstRowAsArray( ['uid' => $uid, 'cid' => $cid]);
```
```suggestion
		$fields = $this->_selectFirstRowAsArray(['uid' => $uid, 'cid' => $cid]);
```

MrPetovan commented on src/Content/Conversation/Repository/UserDefinedChannel.php:
```diff
@@ -62,12 +65,13 @@ public function select(array $condition, array $params = []): UserDefinedChannel
+	public function selectById(int $id, int $uid): UserDefinedChannelEntity
 	{
-		return $this->_selectOne(['id' => $id, 'uid' => $uid]);
+		$fields = $this->_selectFirstRowAsArray( ['id' => $id, 'uid' => $uid]);
```
```suggestion
		$fields = $this->_selectFirstRowAsArray(['id' => $id, 'uid' => $uid]);
```

MrPetovan commented on src/Content/Post/Repository/PostMedia.php:
```diff
@@ -42,17 +44,19 @@ protected function _select(array $condition, array $params = []): BaseCollection
+	public function selectOneById(int $postMediaId): PostMediaEntity
 	{
-		return $this->_selectOne(['id' => $postMediaId]);
+		$fields = $this->_selectFirstRowAsArray( ['id' => $postMediaId]);
```
```suggestion
		$fields = $this->_selectFirstRowAsArray(['id' => $postMediaId]);
```

MrPetovan commented on src/Core/Worker/Repository/Process.php:
```diff
@@ -64,7 +59,9 @@ public function create(int $pid, string $command): Entity\Process
 			}
 
-			$result = $this->_selectOne(['pid' => $pid, 'hostname' => $this->currentHost]);
+			$fields = $this->_selectFirstRowAsArray( ['pid' => $pid, 'hostname' => $this->currentHost]);
```
```suggestion
			$fields = $this->_selectFirstRowAsArray(['pid' => $pid, 'hostname' => $this->currentHost]);
```

MrPetovan reviewed (changes requested):
Tiny formatting issue.

Art4 committed:
Update src/Contact/FriendSuggest/Repository/FriendSuggest.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
Update src/Contact/LocalRelationship/Repository/LocalRelationship.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
Update src/Content/Conversation/Repository/UserDefinedChannel.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
Update src/Content/Post/Repository/PostMedia.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
Update src/Core/Worker/Repository/Process.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

Art4 committed:
let ATProtocol Processor::fetchUriId() return int

Art4 committed:
Merge branch 'develop' into phpstan-level-3

Art4 committed:
Fix code style

Art4 commented:
I fixed the code style. Now we are waiting for the merge of https://git.friendi.ca/friendica/friendica-addons/pulls/1603 and the `2025.02-rc` branch.

Art4 committed:
Merge branch 'develop' into phpstan-level-3

nupplaphil commented:
Hi @Art4  :)

Plz fix:
------ ------------------------------------------------ 
  Line   src/Module/Photo.php                            
 ------ ------------------------------------------------ 
  327    Offset 'nurl' does not exist on array{}|false.  
         🪪  offsetAccess.notFound                       
 ------ ------------------------------------------------ 


There was 1 error:

1) Friendica\Test\Unit\Core\Logger\Factory\SyslogLoggerFactoryTest::testCreateLoggerReturnsPsrLogger
TypeError: Friendica\Core\Logger\Type\SyslogLogger::__construct(): Argument #3 ($logLevel) must be of type int, string given, called in /woodpecker/src/github.com/friendica/friendica/src/Core/Logger/Factory/SyslogLoggerFactory.php on line 58

/woodpecker/src/github.com/friendica/friendica/src/Core/Logger/Type/SyslogLogger.php:89
/woodpecker/src/github.com/friendica/friendica/src/Core/Logger/Factory/SyslogLoggerFactory.php:58
/woodpecker/src/github.com/friendica/friendica/tests/Unit/Core/Logger/Factory/SyslogLoggerFactoryTest.php:39

Art4 committed:
Merge branch 'develop' into phpstan-level-3

Art4 committed:
Fix PHPStan errors

Art4 commented:
@nupplaphil Done. 😊
